smartsim/_core/launcher/dragon/dragonBackend.py

```diff
@@ -0,0 +1,155 @@
+)
+from smartsim.status import (
+    STATUS_COMPLETED,
+    STATUS_FAILED,
+    STATUS_NEVER_STARTED,
+    STATUS_RUNNING,
+)
+
+# pylint: enable=import-error
+
+
+class DragonBackend:
+    """The DragonBackend class is the main interface between
+    SmartSim and Dragon. It is not intended to be user-facing,
+    and will only be called by the Dragon entry-point script or
+    by threads spawned by it.
+    """
+
+    def __init__(self) -> None:
+        self._proc_groups: t.Dict[str, t.Tuple[ProcessGroup, t.List[int]]] = {}
+        self._step_id_lock = RLock()
+        self._step_id = 0
+
+    def _get_new_id(self) -> str:
+        with self._step_id_lock:
+            self._step_id += 1
+            return str(self._step_id)
+
+    @functools.singledispatchmethod
+    # Deliberately suppressing errors so that overloads have the same signature
+    # pylint: disable-next=no-self-use
+    def process_request(self, request: DragonRequest) -> DragonResponse:
+        raise TypeError("Unsure how to process a `{type(request)}` request")
+
+    @process_request.register
+    def _(self, request: DragonRunRequest) -> DragonRunResponse:
+
+        proc = TemplateProcess(
+            target=request.exe,
+            args=request.exe_args,
+            cwd=request.path,
+            env={**request.current_env, **request.env},
+            # stdout=Popen.PIPE,
+            # stderr=Popen.PIPE,
+        )
+
+        grp = ProcessGroup(restart=False, pmi_enabled=request.pmi_enabled)
+        grp.add_process(nproc=request.tasks, template=proc)
+        step_id = self._get_new_id()
+        grp.init()
+        grp.start()
+        self._proc_groups[step_id] = (grp, grp.puids)
+
+        return DragonRunResponse(step_id=step_id)
+
+    @process_request.register
+    def _(self, request: DragonUpdateStatusRequest) -> DragonUpdateStatusResponse:
+        updated_statuses: t.Dict[str, t.Tuple[str, t.Optional[t.List[int]]]] = {}
+        for step_id in request.step_ids:
+            return_codes: t.List[int] = []
+            if step_id in self._proc_groups:
+                proc_group_tuple = self._proc_groups[step_id]
+                if proc_group_tuple[0].status == "Running":
+                    updated_statuses[step_id] = (STATUS_RUNNING, return_codes)
+                else:
+                    if all(proc_id is not None for proc_id in proc_group_tuple[1]):
+                        try:
+                            return_codes = [
+                                Process(None, ident=puid).returncode
+                                for puid in proc_group_tuple[1]
+                            ]
+                        except (ValueError, TypeError):
+                            return_codes = [-1 for _ in proc_group_tuple[1]]
+                    else:
+                        return_codes = [0]
+                    status = (
+                        STATUS_FAILED
+                        if any(return_codes) or proc_group_tuple[0].status == "Error"
```

Comment:
can we avoid case-sensitive exact string matches to be more resilient to changes in the proc_group_tuple content? Maybe using a status enum here `proc_group_tuple[0].status == STATUS_ERROR`?

Reply:
OK, done following the Dragon standards, as part of https://github.com/al-rigazzi/SmartSim/pull/7.